### ISSUE TITLE
Add support for parameters when specifying tweener transition type with "set_trans"

### ISSUE THREAD

**Issue by passivestar** [godotengine/godot-proposals#7802]
### Describe the project you are working on

A sandbox game with the ability to dynamically spawn objects

### Describe the problem or limitation you are having in your project

With transitions like `TRANS_BACK`, `TRANS_ELASTIC` or `TRANS_BOUNCE` you only get one specific shape of a curve without any possibility of customization

This affects any project that needs tweening when you need to add some juice to your object animations. These transitions are also very useful when animating UI elements. In my case, one of the things that I was trying to achieve when porting a Unity project to Godot is this elastic animation when placing objects:

https://github.com/godotengine/godot-proposals/assets/60579014/ac10a0e9-8201-46ab-bfca-552e2f21321f

This is done in Unity with DOTween, and a lot of people seemed to like it. In code it looks like this, where the second parameter to `SetEase` is overshoot:
```csharp
transform.DOScale(scale, .5f).SetEase(Ease.OutElastic, 2f);
```

In Godot it's a little bit milder, but most importantly there's no control over the period of the curve (the number of times it will bounce around before settling to the final value)

https://github.com/godotengine/godot-proposals/assets/60579014/922d0193-337e-4b12-9762-002677e40bf1

While`TRANS_ELASTIC` will be more useful in stylized/cartoonish kind of projects, `TRANS_BACK` on the other hand is much more common, so it's important to be able to customize its overshoot. Not only you usually want a certain amount of overshoot for how you want your project as a whole to "feel", but often you'll also want to have different amount of overshoot for different individual objects and properties. The `TRANS_BACK` curve looks like this right now:

https://github.com/godotengine/godot-proposals/assets/60579014/ce96c015-b08a-4ca3-8fa9-b4545ee78100

The overshoot is very mild

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

The way DOTween does it is by allowing you to pass additional parameters `overshoot`, `amplitude` and `period` into the `SetEase` method. They are then used for different transition types where applicable. This approach seems to be working alright considering the small amount of different transition types that it needs to support

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

Same API can be used for `set_trans`:
```gdscript
var tweener = tween.tween_property(target, "scale", Vector3.ONE * 1.3, 0.2)
tweener.set_trans(Tween.TRANS_BACK, 2.0)
```
Alternatively, if we don't want to touch `set_trans` it can be done with separate methods, which is still much cleaner than the workaround:
```gdscript
tweener.set_trans_overshoot(2.0)
tweener.set_trans_amplitude(2.0)
tweener.set_trans_period(1.0)
```

### If this enhancement will not be used often, can it be worked around with a few lines of script?

A common workaround is to sample a curve:

```gdscript
func tween_step(v, min, max):
    var sampled_value = curve.sample_baked(v)
    var final_value = lerp(min, max, sampled_value)
    target.scale = Vector3.ONE * final_value

func mouse_entered():
    if tween != null:
        tween.kill()
    tween = create_tween()
    tween.tween_method(tween_step.bind(1.0, 1.3), 0.0, 1.0, 1)
```

Not only this is much longer, but it's also very hard to draw a proper back/elastic curve by hand. This is something that's better done with math

### Is there a reason why this should be core and not an add-on in the asset library?

Tweening is widely used in a huge variety of games and apps, that's why there's tweening in godot out of the box in the first place. Making these things configurable will make it massively easier to make nice animations that feel right regardless of what type of project you're working on

**Comment by hubertgdev** [comment]
I'm not an animator, and DOTween's parameters like overshoot have never been clear to me. But I think a very good improvement for Godot tweening would be to use a custom [`Curve`](https://docs.godotengine.org/fr/stable/classes/class_curve.html) instead of using easing and transition presets separately.
Exporting a [`Curve`](https://docs.godotengine.org/fr/stable/classes/class_curve.html) variable also displays a clean curve editor in the inspector. We could even use a preset as default value such as `Curve.Linear` or `Curve.QuadInOut`, that would be pretty neat!

**Comment by passivestar** [comment]
> I'm not an animator, and DOTween's parameters like overshoot have never been clear to me. But I think a very good improvement for Godot tweening would be to use a custom [`Curve`](https://docs.godotengine.org/fr/stable/classes/class_curve.html) instead of using easing and transition presets separately. Exporting a [`Curve`](https://docs.godotengine.org/fr/stable/classes/class_curve.html) variable also displays a clean curve editor in the inspector. We could even use a preset as default value such as `Curve.Linear` or `Curve.QuadInOut`, that would be pretty neat!

Using curves for tweening was already proposed in #1251 and #4304. And I personally think it deserves a second look too. DOTween accepts curves in `SetEase` too, and it can be convenient for some things

However, curves don't solve problems described in this proposal. Even if you could consider a curve an "ok" workaround for custom overshoot of `TRANS_BACK`-like animation, try tweaking the period and amplitude in your Bounce or Elastic transition that you drew:

<img width="300" alt="image" src="https://github.com/godotengine/godot-proposals/assets/60579014/3630bf2c-3399-4f54-804c-6522ebdbad2d">

If these aren't perfect, it will be immediately noticeable that something is off. You need to do these with math

**Comment by Calinou** [comment]
I would prefer just exposing an easing property (as proposed here), rather than allowing the use of arbitrary curves. Creating your own curve is a lot more effort than just tweaking a single value.

**Comment by bitbutter** [comment]
I'm an animator and (sometimes) gamedev. I've used DOTween and godot's new tweens. I echo what @passivestar is saying about the importance of having access to the parameters he mentions vs trying to replicate these effects using curves; very quickly it looks off. the right approach here, imo, is to have the movement generated mathematically, while the user tweaks numbers, previews the result, and repeats to taste.

**Comment by passivestar** [comment]
> I would prefer just exposing an easing property (as proposed here), rather than allowing the use of arbitrary curves. Creating your own curve is a lot more effort than tweaking a single value – this adds up over time if you're trying to iterate on an animation.
> 
> The question is, can we allow elastic/back curves to be adjusted with enough freedom using a single float parameter? Wouldn't you need more than one number to define them?

yeah, like I mentioned DOTween exposed 3 parameters. How they are used depends on the easing type

<img width="600" alt="image" src="https://github.com/godotengine/godot-proposals/assets/60579014/59f25f78-d4e2-47ce-813b-ade2a421504b">

**Comment by KoBeWi** [comment]
Relevant DOTween documentation (see`SetEase()` method): https://dotween.demigiant.com/documentation.php#tweenerSequenceSettings

Adding parameter(s) to `set_trans()` method is not a problem. I'd call them `parm1`, `param2` etc. to not tie them to a specific transition (are they useful for anything else than back/elastic/bounce?). The documentation for each transition type would then mention how they are relevant.

btw Godot also has TRANS_SPRING, you might want to check it.

> Curves

The problem with curves is that GDScript does not support overloads, so it would need to be a new method. Which then brings another problem - setting both transition and curve doesn't make sense. Also Curves can be easily used with `tween_method()`, [the official Tween demo](https://github.com/godotengine/godot-demo-projects/tree/master/2d/tween) even has an example. By "easily" I mean it's a few lines of implementation that can be an addon.

**Comment by passivestar** [comment]
> I'd call them param1, param2 etc. to not tie them to a specific transition

That sounds good to me

> are they useful for anything else than back/elastic/bounce?

I think only those out of existing ones? and spring probably? it's just a little unclear what spring does from the docs, compared to elastic

On a sidenote I want to bring up the fact that DOTween also supports custom easing functions, but I suppose that's not gonna be easy to do for the same reason curves arent easy to add? because of no overloads in GDScript?

**Comment by KoBeWi** [comment]
tbh I'm evaluating how easy it would be to write a general-purpose addon that allows using custom easing functions and turns out it's not exactly possible xd
This is how the demo I mentioned implements curve easing:
```GDScript
tween.tween_method(
	func(v): icon.position = path.position + path.curve.sample_baked(v),
	0.0, path.curve.get_baked_length(), 3.0).set_delay(0.5)
```
It uses lambda method (I put it in separate line for clarity) to assign position and seems like this is the only way. I think for true custom easings we'd need some sort of hybrid Tweener, which would tween a property using values from a method. So the above code would become:
```GDScript
tween.tween_property_method(icon, "position",
	func(v): path.position + path.curve.sample_baked(remap(v, 0, path.curve.get_baked_length()),
	3.0)
```
The point is that lambda becomes self-contained (does not depend on icon position) and can be replaced by an external function:
```GDScript
tween.tween_property_method(icon, "position", interpolate_on_path.bind(path), 3.0)
```
That's a thing for another proposal though. I'd be more in favor of custom functions that allow using curves easily, than using curves directly.

---
Back to this proposal, I can implement it, but I'd need some help with figuring out the equations. All equations can be found here:
https://github.com/godotengine/godot/blob/master/scene/animation/easing_equations.h
Seems like bounce uses the same overshoot parameter as DOTween, so this one is easy:
```C++
namespace back {
static real_t in(real_t t, real_t b, real_t c, real_t d) {
	float s = 1.70158f;
	t /= d;

	return c * t * t * ((s + 1) * t - s) + b;
}
```
Not sure about the rest.

Maybe I'll open a PR that implements extra parameters (without changing the equations) and we'll see.

**Comment by KoBeWi** [comment]
PR up: https://github.com/godotengine/godot/pull/82155

**Comment by timothyqiu** [comment]
> The problem with curves is that GDScript does not support overloads, so it would need to be a new method. Which then brings another problem - setting both transition and curve doesn't make sense.

It could be a separate `set_curve()` method. `set_curve()` and `set_ease()` overwrites each other, i.e., whoever comes last wins, like calling `parallel()` after `chain()`.
